lib/spack/llnl/util/filesystem.py

```diff
@@ -1823,59 +1823,50 @@ def find(root, files, recursive=True):
-        result = _find_recursive(root, files)
-    else:
-        tty.debug(f"Find (not recursive): {root} {str(files)}")
-        result = _find_non_recursive(root, files)
+    if not recursive:
+        max_depth = 0
+
+    tty.debug(f"Find (max depth = {max_depth}): {root} {str(files)}")
+    result = find_max_depth(root, files, max_depth)
 
     tty.debug(f"Find complete: {root} {str(files)}")
     return result
 
 
-@system_path_filter
-def _find_recursive(root, search_files):
-    # The variable here is **on purpose** a defaultdict. The idea is that
-    # we want to poke the filesystem as little as possible, but still maintain
-    # stability in the order of the answer. Thus we are recording each library
-    # found in a key, and reconstructing the stable order later.
-    found_files = collections.defaultdict(list)
-
-    # Make the path absolute to have os.walk also return an absolute path
-    root = os.path.abspath(root)
-    for path, _, list_files in os.walk(root):
-        for search_file in search_files:
-            matches = glob.glob(os.path.join(path, search_file))
-            matches = [os.path.join(path, x) for x in matches]
-            found_files[search_file].extend(matches)
+@system_path_filter(arg_slice=slice(1))
+def find_max_depth(root, globs, max_depth=None):
+    """Given a set of non-recursive glob file patterns, finds all
+    files matching those patterns up to a maximum specified depth.
 
-    answer = []
-    for search_file in search_files:
-        answer.extend(found_files[search_file])
+    Does not search in directories below the specified depth.
 
-    return answer
+    If ``globs`` is a list, files matching earlier entries are placed
+    in the return value before files matching later entries.
+    """
+    if not os.path.exists(root):
+        return []
 
+    root = root
+    if isinstance(globs, str):
+        globs = [globs]
+    regexes = [re.compile(fnmatch.translate(x)) for x in globs]
 
-@system_path_filter
-def _find_non_recursive(root, search_files):
-    # The variable here is **on purpose** a defaultdict as os.list_dir
-    # can return files in any order (does not preserve stability)
     found_files = collections.defaultdict(list)
 
-    # Make the path absolute to have absolute path returned
-    root = os.path.abspath(root)
-
-    for search_file in search_files:
-        matches = glob.glob(os.path.join(root, search_file))
-        matches = [os.path.join(root, x) for x in matches]
-        found_files[search_file].extend(matches)
-
-    answer = []
-    for search_file in search_files:
-        answer.extend(found_files[search_file])
+    dir_queue = collections.deque([(0, root)])
+    while dir_queue:
+        depth, next_dir = dir_queue.pop()
+        for dir_entry in os.scandir(next_dir):
+            if dir_entry.is_dir():
+                if (max_depth is None) or depth < max_depth:
+                    dir_queue.appendleft((depth + 1, os.path.join(next_dir, dir_entry.name)))
```

Comment:
Why not append? `appendleft` is I think linear time, making the thing as a whole quadratic time.

Reply:
I wanted to be able to print a warning for fully-recursive searches beyond some fixed depth. I added a comment about it to the diff. If that doesn't appeal, it can be BFS or DFS (as you say, it is exhaustive).